### PR TITLE
OnlineDDL: Revert for VReplication based migrations

### PR DESCRIPTION
**Pull request by shlomi-noach**
# Ready for review

This PR offers lossless _revert_ for online DDL. Revert is supported for:

### CREATE TABLE and DROP TABLE using any online `ddl_strategy` (either `online`, `gh-ost`, `pt-osc`):

- `CREATE TABLE` : the revert is to `DROP`, but actually implemented by `RENAME...`
- `CREATE TABLE IF NOT EXISTS`: if the `CREATE` was a _noop_, then so is the revert
- `DROP TABLE`: online DDL implements `DRPO` via `RENANE`. Revert "creates" the table by renaming it back into place
- `DROP TABLE IF EXISTS`: if the `DROP` was a _noop_ then so is the revert

### ALTER TABLE using `ddl_strategy=online`

Only VReplication-based `ALTER` migrations are revertible.

### Any reverted migration

It is possible to revert reverted migration, reverted-reverted migrations, and it's turtles all the way down.
 
### Eligible migrations

On top of the above constraints, you may only revert a migration if:

- It is the last successful migration to run on a table.

At this time we only support _revert_ for the last migration, which means you cannot revert the migrations 2nd before last, and you can't "pop" the stack of migrations. However, notice that when you revert a migraiton, the revert itself is a (hopefully) succesful migration, and you may revert the revert.

# What's in a revert?

A revert is an online schema migration of its own. You may submit a revert via `vtctl OnlineDDL <keyspace> revert <uuid>`. This generates a new UUID, which is the job ID for the new revert migration.

_Revert_ migrations are queued and scheduled same as any other migration. They can be tracked, cancelled and retried same as any other migration.

A _revert_ migration is indicated by a SQL statement of the form `revert fb697374_7b25_11eb_85f4_f875a4d24e90`. This statement is internal and not supported in `VTGate` or in any MySQL connection. Work in progress for specialized SQL syntax.

It's important to note that a _revert_ migration does not modify the _entry_ for the migration it reverts. It reads the details of the reverted migration from `_vt.schema_migrations`, and, since it is a migration in itself, has its own entry in that table.

# Implementation

The _revert_ implementation for `CREATE` and `DROP` is relatively simple, in the form of counter-queries that revert the effects of the original migraiton. A revert never actually `CREATE`s a table, and never actually `DROP`s a table. It's implemented behind the scenes by `RENAME`s. There is special logic to handle the case for `CREATE TABLE IF NOT EXISTS` and `DROP TABLE IF EXISTS`.

The implementation for `ALTER`, with `ddl_strategy='online'` is more elaborate, and includes:

- Finding the artifact table from the to-be-reverted `ALTER` migration
- Finding the `_vt.vreplication` entry from the to-be-reverted `ALTER` migration
- Sanity checks
- Re-evaluate a VReplication _rule/filter_ for the new migration
- Read the `pos` from the to-be-reverted VReplication stream. That `pos` was taken while tables were swapped and while no writes took place on the table.
- Construct a new VReplication stream from the above.
- Run the new migration.

We use the `artifacts` value in `_vt.schema_migration` to identify what the original table was.

# Lossless

The user has 24 hours to revert a migration without losing data.

- When you revert a `DROP`, the table reappears with all data intact.
- When you revert an `ALTER`, VReplication makes sure to apply any changes made to the table following the migration, so that you do not lose data.
  - This assumes the new table did not receive changes which are irrevertible. For example, you may have altered a `VARCHAR(10)` to `VARCHAR(20)`. If the new table was populated with now longer texts, then it is irrevertible.
  - Revert-the-revert, revert-the-revert-the-revert and so forth are likewise lossless.

# Tests

Online DDL Revert is tested via a specialize `endtoend` test. See below to get a sense of what gets tested:

```
--- PASS: TestSchemaChange (525.06s)
    --- PASS: TestSchemaChange/CREATE_TABLE_IF_NOT_EXISTS_where_table_does_not_exist (20.04s)
    --- PASS: TestSchemaChange/revert_CREATE_TABLE_IF_NOT_EXISTS_where_did_not_exist (20.05s)
    --- PASS: TestSchemaChange/revert_revert_CREATE_TABLE_IF_NOT_EXISTS_where_did_not_exist (20.05s)
    --- PASS: TestSchemaChange/revert_revert_revert_CREATE_TABLE_IF_NOT_EXISTS_where_did_not_exist (20.05s)
    --- PASS: TestSchemaChange/online_CREATE_TABLE (23.55s)
    --- PASS: TestSchemaChange/revert_CREATE_TABLE (20.05s)
    --- PASS: TestSchemaChange/revert_revert_CREATE_TABLE (20.05s)
    --- PASS: TestSchemaChange/fail_revert_older_change (20.05s)
    --- PASS: TestSchemaChange/CREATE_TABLE_IF_NOT_EXISTS_where_table_exists (20.04s)
    --- PASS: TestSchemaChange/revert_CREATE_TABLE_IF_NOT_EXISTS_where_table_existed (20.05s)
    --- PASS: TestSchemaChange/revert_revert_CREATE_TABLE_IF_NOT_EXISTS_where_table_existed (20.06s)
    --- PASS: TestSchemaChange/fail_online_CREATE_TABLE (20.05s)
    --- PASS: TestSchemaChange/online_ALTER_TABLE_0 (20.07s)
    --- PASS: TestSchemaChange/online_ALTER_TABLE_1 (20.07s)
    --- PASS: TestSchemaChange/revert_ALTER_TABLE (20.09s)
    --- PASS: TestSchemaChange/revert_revert_ALTER_TABLE (20.08s)
    --- PASS: TestSchemaChange/revert_revert_revert_ALTER_TABLE (20.09s)
    --- PASS: TestSchemaChange/online_DROP_TABLE (20.04s)
    --- PASS: TestSchemaChange/revert_DROP_TABLE (20.07s)
    --- PASS: TestSchemaChange/revert_revert_DROP_TABLE (20.06s)
    --- PASS: TestSchemaChange/online_DROP_TABLE_IF_EXISTS (20.07s)
    --- PASS: TestSchemaChange/revert_DROP_TABLE_IF_EXISTS (20.08s)
    --- PASS: TestSchemaChange/revert_revert_DROP_TABLE_IF_EXISTS (20.07s)
    --- PASS: TestSchemaChange/revert_revert_revert_DROP_TABLE_IF_EXISTS (20.07s)
    --- PASS: TestSchemaChange/fail_online_DROP_TABLE (20.07s)
    --- PASS: TestSchemaChange/fail_revert_failed_online_DROP_TABLE (20.07s)
PASS
```

### Points of interest for the reviewer

- `go/test/endtoend/onlineddl_revert/onlineddl_revert_test.go` is the `endtoend` test, and should give some clarity on general behavior.
- In `go/vt/vtctl/vtctl.go` (https://github.com/vitessio/vitess/pull/7478/files#diff-ad37010accc5a29c6751d18124328bb882457e8b2527db688b854c81fe23861c), you can see how a `revert <UUID>` is constructed, and sent just as if it were a normal migration.
- Everything that is important is in `go/vt/vttablet/onlineddl/executor.go`, https://github.com/vitessio/vitess/pull/7478/files#diff-ad37010accc5a29c6751d18124328bb882457e8b2527db688b854c81fe23861c
  - `initVreplicationOriginalMigration()`, https://github.com/vitessio/vitess/pull/7478/files#diff-059c9f46e8d270d9c5514ef2b08679035eb0daaa8d95074e34ef43a81d50dc37R584 is where we init a vreplication DDL for a new `ALTER` (this code existsed before but now refactored into a separate function), and:
  - `initVreplicationRevertMigration()`, https://github.com/vitessio/vitess/pull/7478/files#diff-059c9f46e8d270d9c5514ef2b08679035eb0daaa8d95074e34ef43a81d50dc37R605 is where we inita vreplication DDL which is a _revert_ for a previous VReplication migration. This is where we read the details of the soon-to-be-reverted migration, and speicifcally table & pos.
  - `validateMigrationRevertible()`, https://github.com/vitessio/vitess/pull/7478/files#diff-059c9f46e8d270d9c5514ef2b08679035eb0daaa8d95074e34ef43a81d50dc37R1310, is th elogic where we verify the migration we revert is the last successful one on the given table, etc.
  - `executeRevert()`, https://github.com/vitessio/vitess/pull/7478/files#diff-059c9f46e8d270d9c5514ef2b08679035eb0daaa8d95074e34ef43a81d50dc37R1376, is where all _reverts_ happen. For `ALTER`, the function quickly delegates to vreplication-specific handling. For `CREATE` and `DROP` the logic is within this function. There's a lot of logic to handle edge cases, `IF EXISTS`, `IF NOT EXISTS` etc.
  
---

# Original comment, when this PR was just WIP:

This PR extends #7419 . It only makes sense to review it after #7419 is reviewed & merged.

This **experimental** PR introduces _revert_ for Online DDL via VReplicaiton. With this PR it is possible to revert a successfully completed DDL performed by VReplication on a table (where this was tha last migration executed on said table). The revert operation may take place for as long as the old table is still available, which is now set to `24h`.

This is still evolving, more writeup coming later.

## Related Issue(s)

- extends #7419 
- tracking: #6926 

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [ ] Documentation was added or is not required

## Impacted Areas in Vitess
Components that this PR will affect:

- [ ]  Query Serving
- [x]  VReplication
- [x]  Cluster Management
- [ ]  Build/CI
- [ ]  VTAdmin

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Convenience link to track changes on top of `vreplication-online-ddl` branch (#7419): https://github.com/planetscale/vitess/compare/vreplication-online-ddl...vreplication-online-ddl-revert

**Comment by shlomi-noach** [comment]
# Ready for review

This PR offers lossless _revert_ for online DDL. Revert is supported for:

### CREATE TABLE and DROP TABLE using any online `ddl_strategy` (either `online`, `gh-ost`, `pt-osc`):

- `CREATE TABLE` : the revert is to `DROP`, but actually implemented by `RENAME...`
- `CREATE TABLE IF NOT EXISTS`: if the `CREATE` was a _noop_, then so is the revert
- `DROP TABLE`: online DDL implements `DRPO` via `RENANE`. Revert "creates" the table by renaming it back into place
- `DROP TABLE IF EXISTS`: if the `DROP` was a _noop_ then so is the revert

### ALTER TABLE using `ddl_strategy=online`

Only VReplication-based `ALTER` migrations are revertible.

### Any reverted migration

It is possible to revert reverted migration, reverted-reverted migrations, and it's turtles all the way down.
 
### Eligible migrations

On top of the above constraints, you may only revert a migration if:

- It is the last successful migration to run on a table.

At this time we only support _revert_ for the last migration, which means you cannot revert the migrations 2nd before last, and you can't "pop" the stack of migrations. However, notice that when you revert a migraiton, the revert itself is a (hopefully) succesful migration, and you may revert the revert.

# What's in a revert?

A revert is an online schema migration of its own. You may submit a revert via `vtctl OnlineDDL <keyspace> revert <uuid>`. This generates a new UUID, which is the job ID for the new revert migration.

_Revert_ migrations are queued and scheduled same as any other migration. They can be tracked, cancelled and retried same as any other migration.

A _revert_ migration is indicated by a SQL statement of the form `revert fb697374_7b25_11eb_85f4_f875a4d24e90`. This statement is internal and not supported in `VTGate` or in any MySQL connection. Work in progress for specialized SQL syntax.

It's important to note that a _revert_ migration does not modify the _entry_ for the migration it reverts. It reads the details of the reverted migration from `_vt.schema_migrations`, and, since it is a migration in itself, has its own entry in that table.

# Implementation

The _revert_ implementation for `CREATE` and `DROP` is relatively simple, in the form of counter-queries that revert the effects of the original migraiton. A revert never actually `CREATE`s a table, and never actually `DROP`s a table. It's implemented behind the scenes by `RENAME`s. There is special logic to handle the case for `CREATE TABLE IF NOT EXISTS` and `DROP TABLE IF EXISTS`.

The implementation for `ALTER`, with `ddl_strategy='online'` is more elaborate, and includes:

- Finding the artifact table from the to-be-reverted `ALTER` migration
- Finding the `_vt.vreplication` entry from the to-be-reverted `ALTER` migration
- Sanity checks
- Re-evaluate a VReplication _rule/filter_ for the new migration
- Read the `pos` from the to-be-reverted VReplication stream. That `pos` was taken while tables were swapped and while no writes took place on the table.
- Construct a new VReplication stream from the above.
- Run the new migration.

We use the `artifacts` value in `_vt.schema_migration` to identify what the original table was.

# Lossless

The user has 24 hours to revert a migration without losing data.

- When you revert a `DROP`, the table reappears with all data intact.
- When you revert an `ALTER`, VReplication makes sure to apply any changes made to the table following the migration, so that you do not lose data.
  - This assumes the new table did not receive changes which are irrevertible. For example, you may have altered a `VARCHAR(10)` to `VARCHAR(20)`. If the new table was populated with now longer texts, then it is irrevertible.
  - Revert-the-revert, revert-the-revert-the-revert and so forth are likewise lossless.

# Tests

Online DDL Revert is tested via a specialize `endtoend` test. See below to get a sense of what gets tested:


```
--- PASS: TestSchemaChange (525.06s)
    --- PASS: TestSchemaChange/CREATE_TABLE_IF_NOT_EXISTS_where_table_does_not_exist (20.04s)
    --- PASS: TestSchemaChange/revert_CREATE_TABLE_IF_NOT_EXISTS_where_did_not_exist (20.05s)
    --- PASS: TestSchemaChange/revert_revert_CREATE_TABLE_IF_NOT_EXISTS_where_did_not_exist (20.05s)
    --- PASS: TestSchemaChange/revert_revert_revert_CREATE_TABLE_IF_NOT_EXISTS_where_did_not_exist (20.05s)
    --- PASS: TestSchemaChange/online_CREATE_TABLE (23.55s)
    --- PASS: TestSchemaChange/revert_CREATE_TABLE (20.05s)
    --- PASS: TestSchemaChange/revert_revert_CREATE_TABLE (20.05s)
    --- PASS: TestSchemaChange/fail_revert_older_change (20.05s)
    --- PASS: TestSchemaChange/CREATE_TABLE_IF_NOT_EXISTS_where_table_exists (20.04s)
    --- PASS: TestSchemaChange/revert_CREATE_TABLE_IF_NOT_EXISTS_where_table_existed (20.05s)
    --- PASS: TestSchemaChange/revert_revert_CREATE_TABLE_IF_NOT_EXISTS_where_table_existed (20.06s)
    --- PASS: TestSchemaChange/fail_online_CREATE_TABLE (20.05s)
    --- PASS: TestSchemaChange/online_ALTER_TABLE_0 (20.07s)
    --- PASS: TestSchemaChange/online_ALTER_TABLE_1 (20.07s)
    --- PASS: TestSchemaChange/revert_ALTER_TABLE (20.09s)
    --- PASS: TestSchemaChange/revert_revert_ALTER_TABLE (20.08s)
    --- PASS: TestSchemaChange/revert_revert_revert_ALTER_TABLE (20.09s)
    --- PASS: TestSchemaChange/online_DROP_TABLE (20.04s)
    --- PASS: TestSchemaChange/revert_DROP_TABLE (20.07s)
    --- PASS: TestSchemaChange/revert_revert_DROP_TABLE (20.06s)
    --- PASS: TestSchemaChange/online_DROP_TABLE_IF_EXISTS (20.07s)
    --- PASS: TestSchemaChange/revert_DROP_TABLE_IF_EXISTS (20.08s)
    --- PASS: TestSchemaChange/revert_revert_DROP_TABLE_IF_EXISTS (20.07s)
    --- PASS: TestSchemaChange/revert_revert_revert_DROP_TABLE_IF_EXISTS (20.07s)
    --- PASS: TestSchemaChange/fail_online_DROP_TABLE (20.07s)
    --- PASS: TestSchemaChange/fail_revert_failed_online_DROP_TABLE (20.07s)
PASS
```

### Points of interest for the reviewer

- `go/test/endtoend/onlineddl_revert/onlineddl_revert_test.go` is the `endtoend` test, and should give some clarity on general behavior.
- In `go/vt/vtctl/vtctl.go` (https://github.com/vitessio/vitess/pull/7478/files#diff-ad37010accc5a29c6751d18124328bb882457e8b2527db688b854c81fe23861c), you can see how a `revert <UUID>` is constructed, and sent just as if it were a normal migration.
- Everything that is important is in `go/vt/vttablet/onlineddl/executor.go`, https://github.com/vitessio/vitess/pull/7478/files#diff-ad37010accc5a29c6751d18124328bb882457e8b2527db688b854c81fe23861c
  - `initVreplicationOriginalMigration()`, https://github.com/vitessio/vitess/pull/7478/files#diff-059c9f46e8d270d9c5514ef2b08679035eb0daaa8d95074e34ef43a81d50dc37R584 is where we init a vreplication DDL for a new `ALTER` (this code existsed before but now refactored into a separate function), and:
  - `initVreplicationRevertMigration()`, https://github.com/vitessio/vitess/pull/7478/files#diff-059c9f46e8d270d9c5514ef2b08679035eb0daaa8d95074e34ef43a81d50dc37R605 is where we inita vreplication DDL which is a _revert_ for a previous VReplication migration. This is where we read the details of the soon-to-be-reverted migration, and speicifcally table & pos.
  - `validateMigrationRevertible()`, https://github.com/vitessio/vitess/pull/7478/files#diff-059c9f46e8d270d9c5514ef2b08679035eb0daaa8d95074e34ef43a81d50dc37R1310, is th elogic where we verify the migration we revert is the last successful one on the given table, etc.
  - `executeRevert()`, https://github.com/vitessio/vitess/pull/7478/files#diff-059c9f46e8d270d9c5514ef2b08679035eb0daaa8d95074e34ef43a81d50dc37R1376, is where all _reverts_ happen. For `ALTER`, the function quickly delegates to vreplication-specific handling. For `CREATE` and `DROP` the logic is within this function. There's a lot of logic to handle edge cases, `IF EXISTS`, `IF NOT EXISTS` etc.

**Review by GuptaManan100 — APPROVED** [review]
LGTM!! 🚀

**Review by rohit-nayak-ps — APPROVED** [review]
this is brilliant :-)

File: go/vt/schema/online_ddl.go [hunk 1]
```diff
@@ -208,25 +210,41 @@ func (onlineDDL *OnlineDDL) ToJSON() ([]byte, error) {
...
 	case sqlparser.DropDDLAction:
-		return sqlparser.DropStr, nil
+		return action, sqlparser.DropStr, nil
+	}
+	return action, "", fmt.Errorf("Unsupported online DDL action. SQL=%s", onlineDDL.SQL)
+}
+
+// GetRevertUUID works when this migration is a revert for another migration. It returns the UUID
+// fo the reverted migration.
+// The functio nreturns error when this is not a revert migration.
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/schema/online_ddl.go]: fixed

File: go/test/endtoend/onlineddl_revert/onlineddl_revert_test.go [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+// | zone1-0000003884 |     1 | vt_ks        | stress_test | a0638f6b_ec7b_11ea_9bf8_000d3a9b8a9a | online   | 2020-09-01 17:50:40 | 2020-09-01 17:50:41 | complete         |
+// +------------------+-------+--------------+-------------+--------------------------------------+----------+---------------------+---------------------+------------------+
+
+func checkRecentMigrations(t *testing.T, uuid string, expectStatus schema.OnlineDDLStatus) {
+	result, err := clusterInstance.VtctlclientProcess.OnlineDDLShowRecent(keyspaceName)
+	assert.NoError(t, err)
+	fmt.Println("# 'vtctlclient OnlineDDL show recent' output (for debug purposes):")
+	fmt.Println(result)
+	assert.Equal(t, len(clusterInstance.Keyspaces[0].Shards), strings.Count(result, uuid))
+	// We ensure "full word" regexp becuase some column names may conflict
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/test/endtoend/onlineddl_revert/onlineddl_revert_test.go]: fixed

File: go/test/endtoend/onlineddl_revert/onlineddl_revert_test.go [hunk 3]
```diff
@@ -0,0 +1,776 @@
...
+	t.Run("revert revert CREATE TABLE", func(t *testing.T) {
+		// Restore the table. Data should still be in the table!
+		uuid := testRevertMigration(t, uuids[len(uuids)-1])
+		uuids = append(uuids, uuid)
+		checkRecentMigrations(t, uuid, schema.OnlineDDLStatusComplete)
+		checkTable(t, tableName, true)
+		testSelectTableMetrics(t)
+	})
+	t.Run("fail revert older change", func(t *testing.T) {
+		// We shouldn't be able to revert one-before-last succcessfulk migration.
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/test/endtoend/onlineddl_revert/onlineddl_revert_test.go]: fixed

File: go/test/endtoend/onlineddl_revert/onlineddl_revert_test.go [hunk 4]
```diff
@@ -0,0 +1,776 @@
+/*
+Copyright 2019 The Vitess Authors.
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/test/endtoend/onlineddl_revert/onlineddl_revert_test.go]: fixed

**Pull request merged**